src/features/device-logs/lib/backends/loki.ts

```diff
@@ -0,0 +1,302 @@
+import * as _ from 'lodash';
+import * as Bluebird from 'bluebird';
+import { EventEmitter } from 'events';
+
+import {
+	ClientReadableStream,
+	createInsecureCredentials,
+	createOrgIdMetadata,
+	Direction,
+	EntryAdapter,
+	IGrpcClientAsync,
+	promisifyClient,
+	PusherClient,
+	PushRequest,
+	QuerierClient,
+	QueryRequest,
+	QueryResponse,
+	status,
+	StreamAdapter,
+	TailRequest,
+	TailResponse,
+	Timestamp,
+} from 'loki-grpc-client';
+import { errors } from '@balena/pinejs';
+import { LOKI_HOST, LOKI_PORT } from '../../../../lib/config';
+import {
+	DeviceLog,
+	DeviceLogsBackend,
+	LogContext,
+	LogWriteContext,
+	Subscription,
+} from '../struct';
+import { captureException } from '../../../../infra/error-handling';
+
+const { BadRequestError } = errors;
+
+const MIN_BACKOFF = 100;
+const MAX_BACKOFF = 10 * 1000;
+const VERSION = 1;
+
+function createTimestampFromDate(date = new Date()) {
+	const timestamp = new Timestamp();
+	timestamp.fromDate(date);
+	return timestamp;
+}
+
+function convertToNanoseconds(milliseconds: number, nonce: number) {
+	const MS_TO_NANOS = 1000000;
+	const seconds = Math.floor(milliseconds / 1000);
+	const nanos = (milliseconds - seconds * 1000) * MS_TO_NANOS;
+
+	// use nonce to create unique nanosecond component
+	return [seconds, nanos + nonce];
+}
+
+function backoff(fn: (...args: any) => any, retryIf: (err: Error) => boolean) {
+	return async (...args: any) => {
```

Comment:
You can get typings for this with:
```suggestion
function backoff<T extends (...args: any[]) => any>(fn: T, retryIf: (err: Error) => boolean) {
	return async (...args: Parameters<T>): ReturnType<T> => {
```

Reply:
Thanks, my type foo is not that strong yet.